### ISSUE TITLE
[Bug]: non-zero custom status code considered fatal

### ISSUE THREAD

**Issue by Navil** [realm/realm-dart#1087]
### What happened?

Our sentry reports a few fatal errors, when signing in:

> AppException: non-zero custom status code considered fatal, link to server logs: null

We do not get this on our devices, it is visible to us in sentry. We could reproduce it on a GT-I9505 with Android Version 5.0.1. So maybe it is related to old versions of Android?

The error message does not help much to find the issue.

### Repro steps

Sign in with Firebase (Phone Auth)
Sign in Realm (JWT)
Error

### Version

Channel stable, 3.3.10, on Microsoft Windows [Version 10.0.19045.2364], locale de-AT

### What Realm SDK flavor are you using?

MongoDB Atlas (i.e. Sync, auth, functions)

### What type of application is this?

Flutter Application

### Client OS and version

Android 5.0.1

### Code snippets

```
debugPrint('Verifying phone code'); 
try {
      // Sign the user in (or link) with the credential
      firebase.UserCredential firebaseCredential =
          await firebase.FirebaseAuth.instance.signInWithCredential(credential);
       debugPrint('Signed in with Firebase');

      String jwt = await firebaseCredential.user!.getIdToken();
      user = await app.logIn(Credentials.jwt(jwt));
      _verificationId = null;
      notifyListeners();
      debugPrint('Signed in with Realm');
    } catch (error, stackTrace) {
      Sentry.captureException(error, stackTrace: stackTrace);
    }
```

### Stacktrace of the exception/crash you're getting

```shell
AppException: AppException: non-zero custom status code considered fatal, link to server logs: null
```

### Relevant log output

```shell
Verifying phone code
Signed in with Firebase
AppException: AppException: non-zero custom status code considered fatal, link to server logs: null
```

**Comment by nielsenko** [comment]
How does the jwt look?

**Comment by Navil** [comment]
I do not have those tokens in my logs so I have no token from the crashing device. I could however try to reproduce it on the development environment or provide a working jwt if this helps.

**Comment by nielsenko** [comment]
It might help. Remember to strip the signature part before posting the JWTs here.

I'm also interested in how you have setup Audience (aud claim) and Verification Method on the backend (https://www.mongodb.com/docs/atlas/app-services/authentication/custom-jwt/)

**Comment by Navil** [comment]
So, I can reproduce the issue with an emulator running on 5.0.1.

Here is the content of the jwt:
```
{
  "alg": "RS256",
  "kid": "ce9b880f8182dda557f7c70e20e4e370e3d12747",
  "typ": "JWT"
}
```

```
{
  "iss": "https://securetoken.google.com/myapp-dev",
  "aud": "myapp-dev",
  "auth_time": 1672772941,
  "user_id": "gmeTrLyWFOYGFZxKq1yeLZG8xZt1",
  "sub": "gmeTrLyWFOYGFZxKq1yeLZG8xZt1",
  "iat": 1672772941,
  "exp": 1672776541,
  "phone_number": "+16505550100",
  "firebase": {
    "identities": {
      "phone": [
        "+16505550100"
      ]
    },
    "sign_in_provider": "phone"
  }
}
```

![image](https://user-images.githubusercontent.com/3436399/210425628-62df1f3c-96dc-4701-a659-5b0b1ec713e2.png)


For safety reasons I changed the app name, as it is already on production. As said, in works on every modern device we have, and I can reproduce it on a Pixel 5 emulator with API level 22.

Is this information sufficient?

**Comment by nielsenko** [comment]
Thanks for the detailed information. This is enough for now. I will get back when I know more

**Comment by nielsenko** [comment]
This is more sinister than as such. It has to do with how older versions of android handles TLS certificates. The certificates from https://realm.mongodb.com are signed by lets-encrypt, which is dependent on the root certificate _ISRG Root X1_.

The list of platform that support the _ISRG Root X1_ root certificate can be seen here: https://letsencrypt.org/docs/certificate-compatibility/. In particular you need Android >= 7.1.1

There is a dirty work-around to make these certificates work with old Android apps (>= 2.3.6), that you can read about here: https://letsencrypt.org/docs/dst-root-ca-x3-expiration-september-2021/. Unfortunately I don't think this will work with BoringSSL that is used by Dart / Flutter.

Basically you need configure the http client used to trust the root certificate. It is getting a bit late here, so this is just from the top of my head.
```dart
  final context = SecurityContext();
  context.setTrustedCertificatesBytes(isrgRootX1Cert);
  final client = HttpClient(context: context);
  final appConfig = AppConfig(appId, httpClient: client);
  ...
```

**Comment by Navil** [comment]
Thank your for this throughout information. Strange that no one ran this on an Android verison older than 7.1.1 before apparently.
Is this something you will fix on the end of realm, or shall I build the workaround in my app?

**Comment by nielsenko** [comment]
@Navil Yes, I will pin the certificate in the default http client we create, so you don't have to yourself. But the next release is held up by some regressions in realm-core 13, so you might want to do the fix on your side until then.

**Comment by Navil** [comment]
Thank you very much for the fast help. 
I just tested and can confirm, that it works with that change.

**Comment by nielsenko** [comment]
@desistefanova Pointed out that this is actually documented here: https://www.mongodb.com/docs/realm/sdk/flutter/app-services/connect-to-app/#connect-using-android-7-or-older. I wasn't aware myself.

There is some discussions wrt #1088, if it should land or not. Since there exists a workaround, so you may have to keep doing this after the next release as well.

**Comment by blagoev** [comment]
closing this since there is a workaround and we have #1088 which is in discussion.
